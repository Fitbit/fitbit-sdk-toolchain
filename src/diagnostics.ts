import { relative } from 'path';

import chalk from 'chalk';
import dateformat from 'dateformat';
import indentString from 'indent-string';

interface LineAndCharacter {
  line: number;
  character: number;
}

export interface DiagnosticMessage {
  messageText: string;
  category: DiagnosticCategory;
}

export interface Diagnostic {
  messageText: string | DiagnosticMessage[];
  category: DiagnosticCategory;
  target?: DiagnosticTarget;
  file?: {
    path: string;
    position?:
      | {
          start: { line: number };
        }
      | {
          start: LineAndCharacter;
          end?: LineAndCharacter;
        };
  };
}

export enum DiagnosticCategory {
  Warning = 0,
  Error = 1,
  Suggestion = 2,
  Message = 3,
}

export enum DiagnosticTarget {
  App = 'app',
  Companion = 'companion',
  Settings = 'settings',
}

export type DiagnosticHandler = (diagnostic: Diagnostic) => void;

const consoleColors = {
  [DiagnosticCategory.Warning]: chalk.keyword('orange'),
  [DiagnosticCategory.Error]: chalk.red,
  [DiagnosticCategory.Message]: chalk,
  [DiagnosticCategory.Suggestion]: chalk,
};

const categoryNames = {
  [DiagnosticCategory.Warning]: 'warn',
  [DiagnosticCategory.Error]: 'error',
  [DiagnosticCategory.Message]: 'info',
  [DiagnosticCategory.Suggestion]: 'suggestion',
};

export const logDiagnosticToConsole: DiagnosticHandler = (diagnostic) => {
  const { category, messageText, target } = diagnostic;

  let filePosition = '';
  if (diagnostic.file) {
    const { path, position } = diagnostic.file;
    filePosition += ` ${relative(process.cwd(), path)}`;
    if (position) {
      filePosition += `:${position.start.line + 1}`;
      if ('character' in position.start) {
        filePosition += `,${position.start.character + 1}`;
      }
    }
  }

  const now = new Date();
  const timestamp = dateformat(now, 'HH:MM:ss');

  const colorizer = consoleColors[category];
  const categoryName = categoryNames[category];
  const targetString = target ? target : 'build';
  const message =
    typeof messageText === 'string' ? messageText : messageText[0].messageText;

  console.log(
    colorizer(
      `[${timestamp}][${categoryName}][${targetString}]${filePosition} ${message}`,
    ),
  );

  if (typeof messageText !== 'string') {
    for (let i = 1; i < messageText.length; i += 1) {
      console.log(
        consoleColors[messageText[i].category](
          indentString(messageText[i].messageText, 4),
        ),
      );
    }
  }
};
