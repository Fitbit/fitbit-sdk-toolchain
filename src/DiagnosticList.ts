import { Diagnostic, DiagnosticCategory } from './diagnostics';

export default class DiagnosticList {
  fatalError = false;

  constructor(public diagnostics = [] as Diagnostic[]) {
    if (!Array.isArray(diagnostics)) {
      throw new TypeError(
        `diagnostics must be array, not ${typeof diagnostics}`,
      );
    }
  }

  extend(otherDiagnosticList: DiagnosticList) {
    // Add "target" and timestamp iff (if and only if) they don't already exist.
    this.diagnostics.push(...otherDiagnosticList.diagnostics);
    if (otherDiagnosticList.fatalError) this.fatalError = true;
  }

  push(diagnostic: Diagnostic) {
    if (
      diagnostic.messageText === undefined ||
      diagnostic.category === undefined
    ) {
      throw new TypeError(
        'Diagnostic is missing `messageText` or `category` keys',
      );
    }
    this.diagnostics.push(diagnostic);
  }

  pushMessage(text: string) {
    this.push({
      messageText: text,
      category: DiagnosticCategory.Message,
    });
  }

  pushWarning(text: string) {
    this.push({
      messageText: text,
      category: DiagnosticCategory.Warning,
    });
  }

  pushError(text: string) {
    this.push({
      messageText: text,
      category: DiagnosticCategory.Error,
    });
  }

  pushFatalError(text: string) {
    this.pushError(text);
    this.fatalError = true;
  }
}
