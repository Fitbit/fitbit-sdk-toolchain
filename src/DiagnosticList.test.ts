import { DiagnosticCategory, Diagnostic } from './diagnostics';
import DiagnosticList from './DiagnosticList';

it('pushes a message to the diagnostic list', () => {
    const list = new DiagnosticList();
    list.pushMessage('I am a message');
    expect(list.diagnostics).toEqual([
        expect.objectContaining({
            messageText: 'I am a message',
            category: DiagnosticCategory.Message,
        }),
    ]);
    expect(list.fatalError).toEqual(false);
});

it('pushes a warning to the diagnostic list', () => {
    const list = new DiagnosticList();
    list.pushWarning('I am a warning');
    expect(list.diagnostics).toEqual([
        expect.objectContaining({
            messageText: 'I am a warning',
            category: DiagnosticCategory.Warning,
        }),
    ]);
    expect(list.fatalError).toEqual(false);
});

it('pushes an error to the diagnostic list', () => {
    const list = new DiagnosticList();
    list.pushError('I am error');
    expect(list.diagnostics).toEqual([
        expect.objectContaining({
            messageText: 'I am error',
            category: DiagnosticCategory.Error,
        }),
    ]);
    expect(list.fatalError).toEqual(false);
});

it('pushes a fatal error to the diagnostic list', () => {
    const list = new DiagnosticList();
    list.pushFatalError('I ded');
    expect(list.diagnostics).toEqual([
        expect.objectContaining({
            messageText: 'I ded',
            category: DiagnosticCategory.Error,
        }),
    ]);
    expect(list.fatalError).toEqual(true);
});

it('pushes a Diagnostic directly to the list', () => {
    const list = new DiagnosticList();
    const diagnostic = {
        messageText: 'Message text',
        category: DiagnosticCategory.Warning,
    };
    list.push(diagnostic);
    expect(list.diagnostics).toEqual([expect.objectContaining(diagnostic)]);
});

it('tries to push a diagnostic to the list without category', () => {
    const list = new DiagnosticList();
    function tryPushBadMessage(): void {
        list.push({ messageText: 'Message!' } as Diagnostic);
    }
    expect(tryPushBadMessage).toThrowErrorMatchingSnapshot();
});

it('tries to push a diagnostic to the list without messageText', () => {
    const list = new DiagnosticList();
    function tryPushBadMessage(): void {
        list.push({ category: DiagnosticCategory.Error } as Diagnostic);
    }
    expect(tryPushBadMessage).toThrowErrorMatchingSnapshot();
});

it('pushes multiple diagnostics to the list', () => {
    const list = new DiagnosticList();
    list.pushMessage('Message 1');
    list.pushError("I'm sorry Dave, I can't let you do that");
    list.pushMessage('Message 2');
    list.pushWarning('Uh oh');
    expect(list.diagnostics).toEqual([
        expect.objectContaining({
            messageText: 'Message 1',
            category: DiagnosticCategory.Message,
        }),
        expect.objectContaining({
            messageText: "I'm sorry Dave, I can't let you do that",
            category: DiagnosticCategory.Error,
        }),
        expect.objectContaining({
            messageText: 'Message 2',
            category: DiagnosticCategory.Message,
        }),
        expect.objectContaining({
            messageText: 'Uh oh',
            category: DiagnosticCategory.Warning,
        }),
    ]);
    expect(list.fatalError).toEqual(false);
});

it('merges two lists together', () => {
    const listA = new DiagnosticList();
    const listB = new DiagnosticList();
    listA.pushMessage('List A message 1');
    listA.pushMessage('List A message 2');
    listB.pushMessage('List B message 1');
    listA.extend(listB);
    expect(listA.diagnostics).toEqual([
        expect.objectContaining({
            messageText: 'List A message 1',
            category: DiagnosticCategory.Message,
        }),
        expect.objectContaining({
            messageText: 'List A message 2',
            category: DiagnosticCategory.Message,
        }),
        expect.objectContaining({
            messageText: 'List B message 1',
            category: DiagnosticCategory.Message,
        }),
    ]);
    expect(listA.fatalError).toEqual(false);
});

it('merges two empty lists', () => {
    const listA = new DiagnosticList();
    const listB = new DiagnosticList();
    listA.extend(listB);
    expect(listA.diagnostics).toEqual([]);
    expect(listA.fatalError).toEqual(false);
});

it('merges an empty list into a nonempty list', () => {
    const listA = new DiagnosticList();
    const listB = new DiagnosticList();
    listA.pushMessage('A message');
    listA.extend(listB);
    expect(listA.diagnostics).toEqual([
        expect.objectContaining({
            messageText: 'A message',
            category: DiagnosticCategory.Message,
        }),
    ]);
    expect(listA.fatalError).toEqual(false);
});

it('merges a nonempty list into an empty list', () => {
    const listA = new DiagnosticList();
    const listB = new DiagnosticList();
    listB.pushMessage('A message');
    listA.extend(listB);
    expect(listA.diagnostics).toEqual([
        expect.objectContaining({
            messageText: 'A message',
            category: DiagnosticCategory.Message,
        }),
    ]);
    expect(listA.fatalError).toEqual(false);
});

it('merges a list with a fatal error', () => {
    const listA = new DiagnosticList();
    const listB = new DiagnosticList();
    listB.pushFatalError('Blarg im ded');
    listA.extend(listB);
    expect(listA.fatalError).toEqual(true);
});
