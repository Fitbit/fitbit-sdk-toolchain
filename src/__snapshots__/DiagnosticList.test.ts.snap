// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tries to push a diagnostic to the list without category 1`] = `"Diagnostic is missing \`messageText\` or \`category\` keys"`;

exports[`tries to push a diagnostic to the list without messageText 1`] = `"Diagnostic is missing \`messageText\` or \`category\` keys"`;
